File: lua/ssns/adapters/factory.lua
---@class AdapterFactory
local AdapterFactory = {}

---Supported database types
---@type table<string, boolean>
local SUPPORTED_TYPES = {
  sqlserver = true,
  postgres = true,
  mysql = true,
  sqlite = true,
}

---Create an adapter instance for a specific database type
---@param db_type string The database type (sqlserver, mysql, postgres, sqlite)
---@param connection_config ConnectionData The connection configuration
---@return BaseAdapter? adapter The created adapter instance or nil if type unknown
---@return string? error_message Error message if adapter creation failed
function AdapterFactory.create_adapter_for_type(db_type, connection_config)
  if not db_type or db_type == "" then
    return nil, "Database type is empty"
  end

  if not SUPPORTED_TYPES[db_type] then
    return nil, string.format("Unsupported database type: %s", db_type)
  end

  -- Load the appropriate adapter module
  local adapter_module_name = string.format("ssns.adapters.%s", db_type)
  local ok, adapter_module = pcall(require, adapter_module_name)

  if not ok then
    return nil, string.format("Failed to load adapter for %s: %s", db_type, adapter_module)
  end

  -- Create and return the adapter instance with connection config
  local adapter = adapter_module.new(connection_config)
  return adapter, nil
end

---Get list of supported database types
---@return string[] db_types Array of supported database type identifiers
function AdapterFactory.get_supported_types()
  local types = {}

  for db_type, _ in pairs(SUPPORTED_TYPES) do
    table.insert(types, db_type)
  end

  table.sort(types)
  return types
end

---Check if a database type is supported
---@param db_type string
---@return boolean
function AdapterFactory.is_supported(db_type)
  return SUPPORTED_TYPES[db_type] == true
end

---Validate that an adapter module exists for a database type
---@param db_type string
---@return boolean exists
function AdapterFactory.adapter_exists(db_type)
  local adapter_module_name = string.format("ssns.adapters.%s", db_type)
  local ok, _ = pcall(require, adapter_module_name)
  return ok
end

---Register a custom database type
---Allows users to add support for custom database types
---@param db_type string The database type identifier
function AdapterFactory.register_type(db_type)
  SUPPORTED_TYPES[db_type] = true
end

return AdapterFactory

File: rplugin/node/ssns-db/drivers/factory.js
/**
 * Driver Factory - Create the appropriate database driver based on config type
 *
 * Supports:
 * - SQL Server (type: "sqlserver")
 * - MySQL (type: "mysql")
 * - SQLite (type: "sqlite")
 * - PostgreSQL (type: "postgres")
 */
const { ssnsLog } = require('../ssns-log');
const SqlServerDriver = require('./sqlserver');
const MySQLDriver = require('./mysql');
const SQLiteDriver = require('./sqlite');
const PostgresDriver = require('./postgres');

/**
 * Get the appropriate driver for a connection config
 *
 * @param {Object} config - Connection configuration object
 * @param {string} config.type - Database type ("sqlserver", "mysql", "sqlite", "postgres")
 * @param {Object} config.server - Server connection details
 * @param {Object} config.auth - Authentication details
 * @param {Object} [config.options] - Additional connection options
 * @returns {BaseDriver} Driver instance
 * @throws {Error} If config is invalid or type is unsupported
 */
function getDriver(config) {
  ssnsLog(`[factory] getDriver called with config type: ${config && config.type}`);

  if (!config || typeof config !== 'object') {
    ssnsLog('[factory] Invalid config: must be a non-empty object');
    throw new Error('Invalid config: must be a non-empty object');
  }

  if (!config.type) {
    ssnsLog('[factory] Invalid config: missing type field');
    throw new Error('Invalid config: missing type field');
  }

  const dbType = config.type.toLowerCase();

  // SQL Server
  if (dbType === 'sqlserver' || dbType === 'mssql') {
    ssnsLog('[factory] Creating SQL Server driver');
    return new SqlServerDriver(config);
  }

  // MySQL
  if (dbType === 'mysql') {
    ssnsLog('[factory] Creating MySQL driver');
    return new MySQLDriver(config);
  }

  // SQLite
  if (dbType === 'sqlite') {
    ssnsLog('[factory] Creating SQLite driver');
    return new SQLiteDriver(config);
  }

  // PostgreSQL
  if (dbType === 'postgres' || dbType === 'postgresql') {
    ssnsLog('[factory] Creating PostgreSQL driver');
    return new PostgresDriver(config);
  }

  // Unknown database type
  ssnsLog(`[factory] Unsupported database type: ${config.type}`);
  throw new Error(
    `Unsupported database type: ${config.type}\n` +
    `Supported types:\n` +
    `  - sqlserver\n` +
    `  - mysql\n` +
    `  - sqlite\n` +
    `  - postgres`
  );
}

/**
 * Check if a database type is supported
 *
 * @param {string} dbType - Database type to check
 * @returns {boolean} True if supported
 */
function isSupported(dbType) {
  ssnsLog(`[factory] isSupported called with: ${dbType}`);
  const supported = ['sqlserver', 'mssql', 'mysql', 'sqlite', 'postgres', 'postgresql'];
  const result = supported.includes(dbType.toLowerCase());
  ssnsLog(`[factory] isSupported result: ${result}`);
  return result;
}

/**
 * Get list of supported database types
 *
 * @returns {Array<string>} List of supported database types
 */
function getSupportedTypes() {
  ssnsLog('[factory] getSupportedTypes called');
  return ['sqlserver', 'mysql', 'sqlite', 'postgres'];
}

module.exports = {
  getDriver,
  isSupported,
  getSupportedTypes
};
